tensorizer/serialization.py

```diff
@@ -3748,90 +3763,127 @@ def _interrupt_transfer():
+            elif self._encrypted and w.tensor.device.type == "cpu":
+                address = w.tensor.data_ptr()
+                if address in seen_addresses:
+                    to_clone_write_specs.append(w)
+                    result.append(clone_iter_holder)
+                else:
+                    seen_addresses.add(address)
+                    noop_write_specs.append(w)
+                    result.append(noop_iter_holder)
+            else:
+                noop_write_specs.append(w)
+                result.append(noop_iter_holder)
+
+        if cuda_write_specs:
+            (
+                cuda_iter,
+                interrupt_transfer,
+            ) = self._async_bulk_device_to_host_transfer(cuda_write_specs)
+            cuda_result_iter_holder.iter = cuda_iter
+        else:
+            interrupt_transfer = None
+
+        noop_iter_holder.iter = iter(noop_write_specs)
+        non_contiguous_result_iter_holder.iter = self._computation_pool.map(
+            self._make_contiguous, non_contiguous_write_specs
+        )
+        clone_iter_holder.iter = self._computation_pool.map(
+            self._clone_detach, to_clone_write_specs
+        )
+
+        def result_iter():
+            for r in result:
+                yield next(r.iter)
+
+        return result_iter(), interrupt_transfer
 
-    def _bulk_write(self, tensors: Iterable[_WriteSpec]):
-        tensors = collections.deque(tensors)
+    def _bulk_write(self, write_specs: Iterable[_WriteSpec]):
+        write_specs = collections.deque(write_specs)
         next_pos = self._file.tell()
         if _syscalls.has_fallocate() and self._fd:
-            size = sum(len(t.name) for t in tensors)
+            size = sum(len(w.name) for w in write_specs)
             size += sum(
-                t.tensor.element_size()
-                * t.tensor.nelement()
-                * (not t.tensor.is_meta)
-                for t in tensors
+                w.tensor.element_size()
+                * w.tensor.nelement()
+                * (not w.tensor.is_meta)
+                for w in write_specs
             )
             # Rough underestimate of header size
             header_min_size = 24
-            size += header_min_size * len(tensors)
+            size += header_min_size * len(write_specs)
             _syscalls.try_fallocate(
                 self._fd, next_pos, size, suppress_all_errors=True
             )
 
-        cuda_tensors = [
-            t.tensor for t in tensors if t.tensor.device.type == "cuda"
-        ]
-        if cuda_tensors:
-            (
-                transferred,
-                interrupt_transfer,
-            ) = self._async_bulk_device_to_host_transfer(cuda_tensors)
-        else:
-            transferred = interrupt_transfer = None
-        del cuda_tensors
-
-        if self._encrypted:
-            shared = []
-            seen_addresses = set()
-            for t in reversed(tensors):
```

Comment:
@Eta0 I am not carrying forward this `reversed()`. Not sure if that was logic just to keep with the `popleft`-iness, or if there was some need to specifically clone() the earlier tensors instead of the later ones.

Reply:
The `reversed` is for correctness. Earlier tensors need to be cloned instead of later ones because of the following scenario where A and B share memory:

- A is handed to the writer pipeline with its original buffer
- A starts to be encrypted in place in the buffer
- B is cloned into a new buffer by the buffer-preparing pipeline and enqueued while this is happening
  - The state of B is now encrypted or partially-encrypted garbage data
- A finishes being written and decrypts in place
- B is handed to the writer
- B ends up being written double-encrypted

When the earlier tensors are cloned, the original buffer is only given to the writer thread (and opened up to being temporarily modified) *after* the clones have been made, so the clones are guaranteed to have the correct data.

Reply:
For what it's worth this situation also can't come up if all cloning happens before any writing, but ideally we are interleaving CPU-bound operations like cloning with the IO-bound operations like writing to maximize throughput on all resource types at once instead of batching them (which is the reason for the design of the device transfer thread)